src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs

```diff
@@ -243,16 +244,65 @@ impl<'tcx> GotocCtx<'tcx> {
     fn handle_kanitool_attributes(&mut self) {
         let instance = self.current_fn().instance();
 
+        // Vectors for storing instances of each attribute type,
+        // TODO: This can be modifed to use Enums when more options are provided
+        let mut attribute_vector = vec![];
+        let mut proof_attribute_vector = vec![];
+
+        // Loop through instances to get all "kanitool::x" attribute strings
         for attr in self.tcx.get_attrs(instance.def_id()) {
-            match kanitool_attr_name(attr).as_deref() {
-                Some("proof") => self.handle_kanitool_proof(),
-                _ => {}
+            // Get the string the appears after "kanitool::" in each attribute string.
+            // Ex - "proof" | "unwind" etc.
+            if let Some(attribute_string) = kanitool_attr_name(attr).as_deref() {
+                // Push to proof vector
+                if attribute_string == "proof" {
+                    proof_attribute_vector.push(attr);
+                }
+                // Push to attribute vector that can be expanded to a map when more options become available
+                else {
+                    attribute_vector.push((attribute_string.to_string(), attr));
+                }
             }
         }
+
+        // In the case when there's only one proof attribute (correct behavior), create harness and modify it
+        // depending on each subsequent attribute that's being called by the user.
+        if proof_attribute_vector.len() == 1 {
```

Comment:
These cases do not seem to overlap so I am not sure why you are using if-then-else statements here. I would also recommend handling them in small functions if possible.

Reply:
They do overlap as it's necessary to check to how many proof attributes have been declared per harness and depending on the number, there's different handling that needs to happen. The entire function has a single purpose so it didn't seem necessary to factor it out into smaller functions.

Reply:
@jaisnan, if I understand this code correctly, you are trying to restrict any other `kani` attribute to things that are tagged with the `proof` attribute, right?

If that is the case, I would recommend to change this to have something like the following structure:
```rust
        let attrs: kani_attributes(self.tcx.get_attrs(instance.def_id()));  // BTreeMap<&str, &rustc_ast::Attribute>
        if attrs.contains_key("proof") {
            // Handle attributes here.
        } else {
            if attrs.len() > 0 {
                // Warning: We currently only support attributes on proof harnesses. The attribute {} will be ignored.
            }
        }
```

Reply:
I think using a BTreeMap would prevent us from checking if there are duplicate attributes in the same harness or not. If someone calls two `unwind` annotations then, the last one is going to be read and parsed if we use a BTreeMap I think. 

We ideally should not just check if a key exists but also how many instances of them. Maybe I can use a multimap? https://docs.rs/btreemultimap/latest/btreemultimap/ ?